pip_cache/__init__.py

```diff
@@ -66,36 +94,55 @@ def update_package_list():
         import xmlrpclib
     except ImportError:
         import xmlrpc.client as xmlrpclib
-    print('Connecting to PyPi...', end='')
-    sys.stdout.flush()
+
+    pip_cache_data_dir = get_pip_cache_data_dir()
+    raw_index_filename = get_raw_index_filename()
+    index_filename = get_index_filename()
+
+    # make sure that the directory we might be writing in exists
+    if not os.path.isdir(pip_cache_data_dir):
+        os.mkdir(pip_cache_data_dir)
+
+    print('Connecting to PyPi...', end='', flush=True)
```

Comment:
FYA: there's a neat little package, [manrajgrover/halo](https://github.com/manrajgrover/halo), which adds a little spinner in the console and can be used as a context manager

```py
with Halo('Connecting to PyPi'):
    client = xmlrpclib.ServerProxy('https://pypi.python.org/pypi')

with Halo('Downloading package names'):
    packages = client.list_packages()

with Halo('Writing packages to cache'):
    with open(raw_index_filename, 'w') as f:
        f.write('\n'.join(packages))

with Halo('Indexing packages'):
    trie = marisa_trie.Trie(packages)
    trie.save(index_filename)
```

![Peek 2020-05-17 17-20](https://user-images.githubusercontent.com/33840/82160362-c22a7200-9862-11ea-8062-9da77eeb22c3.gif)

I like to save the text of each thing, though

![Peek 2020-05-17 17-21](https://user-images.githubusercontent.com/33840/82160386-f9008800-9862-11ea-800a-5611fc550f61.gif)

```py
class task(Halo):
    def __exit__(self, type, value, traceback):
        self.succeed()

with task('Connecting to PyPi'):
    client = xmlrpclib.ServerProxy('https://pypi.python.org/pypi')

with task('Downloading package names'):
    packages = client.list_packages()

with task('Writing packages to cache'):
    with open(raw_index_filename, 'w') as f:
        f.write('\n'.join(packages))

with task('Indexing packages'):
    trie = marisa_trie.Trie(packages)
    trie.save(index_filename)
```